ouroboros-consensus/src/Ouroboros/Consensus/Node/ErrorPolicy.hs

```diff
@@ -0,0 +1,152 @@
+          -- them, we'd somehow have to distinguish between IO exceptions
+          -- arising from disk I/O (shutdownNode) and those arising from
+          -- network failures (SuspendConsumer).
+          ErrorPolicy $ \(_ :: VolatileDBError)  -> Just shutdownNode
+        , ErrorPolicy $ \(_ :: ChainDbFailure)   -> Just shutdownNode
+        , ErrorPolicy $ \(_ :: FsError)          -> Just shutdownNode
+        , ErrorPolicy $ \(_ :: ImmutableDBError) -> Just shutdownNode
+        , ErrorPolicy $ \(_ :: DbMarkerError)    -> Just shutdownNode
+
+          -- Node configuration failure
+        , ErrorPolicy $ \(_ :: PBftLeaderCredentialsError) -> Just shutdownNode
+
+          -- Some chain DB errors are indicative of a bug in our code, others
+          -- indicate an invalid request from the peer. If the DB is closed
+          -- entirely, it will only be reopened after a node restart.
+        , ErrorPolicy $ \(e :: ChainDbError) ->
+            case e of
+              NoGenesisBlock{}       -> Just theyBuggyOrEvil
+              ClosedDBError{}        -> Just shutdownNode
+              ClosedReaderError{}    -> Just ourBug
+              InvalidIteratorRange{} -> Just theyBuggyOrEvil
+
+          -- We have some resource registries that are used per-connection,
+          -- and so if we have ResourceRegistry related exception, we close
+          -- the connection but leave the rest of the node running.
+        , ErrorPolicy $ \(_ :: RegistryClosedException)         -> Just ourBug
+        , ErrorPolicy $ \(_ :: ResourceRegistryThreadException) -> Just ourBug
+
+          -- An exception in the block fetch server meant the client asked
+          -- for some blocks we used to have but got GCed. This means the
+          -- peer is on a chain that forks off more than @k@ blocks away.
+        , ErrorPolicy $ \(_ :: BlockFetchServerException) -> Just distantPeer
+
+          -- Some chain sync client exceptions indicate malicious behaviour,
+          -- others merely mean that we should disconnect from this client
+          -- because we have diverged too much.
+        , ErrorPolicy $ \(e :: ChainSyncClientException) ->
+            case e of
+              HeaderExceedsClockSkew{} -> Just misconfiguredPeer
+              ForkTooDeep{}            -> Just distantPeer
+              ChainError{}             -> Just theyBuggyOrEvil
+              InvalidRollForward{}     -> Just distantPeer
+              InvalidRollBack{}        -> Just theyBuggyOrEvil
+              InvalidIntersection{}    -> Just theyBuggyOrEvil
+              NoMoreIntersection{}     -> Just distantPeer
+              DoesntFit{}              -> Just theyBuggyOrEvil
+              InvalidBlock{}           -> Just theyBuggyOrEvil
+
+          -- Dispatch on nested exception
+        , ErrorPolicy $ \(ExceptionInLinkedThread _ e) ->
+            evalErrorPolicies e (epAppErrorPolicies consensusErrorPolicy)
+        ]
```

Comment:
I am not familiar with all the exceptions in the `ouroboros-consensus`.  So I grepped, what about:
* `OnSlotException` which can be raised by `onSlot` function
* `BlockFetchServerException` defined in `Ouroboros.Consensus.BlockFetchServer`

Just a note to myself: we also need to take care about `ouroboros-network` exceptions:
* `ClientProtocolError` - thrown as an application exception by the `handshake` protocol
* `TxSubsmissionProtocolError`s (inbound and outbound)
* `BlockFetchProtocolFailure`

Reply:
`OnSlotException` I moved to the tests so it's not a concern. `BlockFetchServerException` we do handle:

```haskell
        , ErrorPolicy $ \(_ :: BlockFetchServerException) -> Just distantPeer
```